x/mongo/driver/topology/server_test.go

```diff
@@ -54,6 +54,217 @@ func (cncd *channelNetConnDialer) DialContext(_ context.Context, _, _ string) (n
+						// Sleep for at least 150ms and expect the context passed to
+						// server.Connection() to time out during the sleep. Expect the error
+						// returned by GetHandshakeInformation() to be treated as a timeout caused
+						// by an operation-scoped deadline.
+						// E.g. FindOne(context.WithTimeout(...))
+						time.Sleep(150 * time.Millisecond)
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: false,
+		},
+		{
+			desc: "Child context deadline exceeded error during handshake should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 100*time.Millisecond)
+			},
+			handshaker: func(Handshaker) Handshaker {
+				h := auth.Handshaker(nil, &auth.HandshakeOptions{})
+				return &testHandshaker{
+					getHandshakeInformation: func(ctx context.Context, addr address.Address, c driver.Connection) (driver.HandshakeInformation, error) {
+						// Wrap the context in a context with an already-exceeded deadline. Expect
+						// the error returned by GetHandshakeInformation() to be treated as a
+						// timeout caused by connectTimeoutMS.
+						ctx, cancel := context.WithDeadline(ctx, time.Now().Add(-1*time.Hour))
+						defer cancel()
+						return h.GetHandshakeInformation(ctx, addr, c)
+					},
+				}
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+		{
+			desc: "Dial errors unrelated to context timeouts should clear the pool",
+			ctxFn: func() (context.Context, context.CancelFunc) {
+				return context.WithTimeout(context.Background(), 1*time.Second)
+			},
+			dialer: func(Dialer) Dialer {
+				var d net.Dialer
+				return DialerFunc(func(ctx context.Context, _, _ string) (net.Conn, error) {
+					// Try to dial an invalid TCP address and expect an error.
+					return d.DialContext(ctx, "tcp", "300.0.0.0:nope")
+				})
+			},
+			expectErr:         true,
+			expectPoolCleared: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.desc, func(t *testing.T) {
+			t.Parallel()
+
+			// Start a goroutine that pulls events from the events channel and inserts them into the
+			// events slice. Use a sync.WaitGroup to allow the test code to block until the events
+			// channel loop exits, guaranteeing that all events were copied from the channel.
+			events := make([]*event.PoolEvent, 0)
+			eventsCh := make(chan *event.PoolEvent)
+			var eventsWg sync.WaitGroup
+			eventsWg.Add(1)
+			go func() {
+				for evt := range eventsCh {
+					events = append(events, evt)
+				}
+				eventsWg.Done()
+			}()
+
+			// Create a TCP listener on a random port but never call Accept on it. The TCP listener will
+			// behave like a TCP service that doesn't explicitly deny connections, but doesn't respond to
+			// them, forcing client-side timeouts instead.
```

Comment:
Test cases like `Parent context deadline exceeded error during handshake should not clear the pool` are overriding the Handshaker, but not the Dialer. Wouldn't they need connection to succeed to perform the handshake? I am not sure how that is possible if this test server is not accepting connections.